File: 2370/solution.java
import java.util.Arrays;

class Solution {
    public int longestIdealString(String s, int k) {
        int[] dp = new int[26];
        for (int i = 0; i < s.length(); i++) {
            int currValue = (int) (s.charAt(i) - 'a');
            int currTotal = 1;
            int bot = Math.max(0, currValue - k);
            int top = Math.min(26, currValue + k + 1);
            for (int j = bot; j < top; j++) {
                if (Math.abs(j - currValue) <= k) {
                    currTotal = Math.max(currTotal, dp[j] + 1);
                }
            }
            dp[currValue] = currTotal;
        }
        return Arrays.stream(dp).max().orElse(0);
    }
}

// Time O(n*26)
// Space O(26)

File: 2370/solution.ts
function longestIdealString(s: string, k: number): number {
  const dp = new Array(26).fill(0);
  for (let i = 0; i < s.length; i++) {
    const currValue = s.charCodeAt(i) - "a".charCodeAt(0);
    const bot = Math.max(0, currValue - k);
    const top = Math.min(26, currValue + k + 1);
    dp[currValue] = Math.max(...dp.slice(bot, top)) + 1;
  }
  return Math.max(...dp);
}
